fms/utils/serialization.py

```diff
@@ -138,12 +226,257 @@ def load_state_dict(
+
+    Args
+    ====
+    param: torch.nn.Parameter
+        Parameter that has had TP applied
+    tensor_value: torch.Tensor
+        tensor that needs sharding
+    rank: int
+        Rank of the current process
+    world_size: int
+        Total number of TP processes
+    """
+    # Divide the weight matrix along the first dimension.
+    output_size_per_partition = param.shape[0]
+    if not is_bias:
+        tensor = tensor_value[
+            (rank * output_size_per_partition) : (
+                (rank + 1) * output_size_per_partition
+            ),
+            :,
+        ]
+    else:
+        tensor = tensor_value[
+            (rank * output_size_per_partition) : (
+                (rank + 1) * output_size_per_partition
+            )
+        ]
+    param.copy_(tensor, non_blocking=True)
+
+
+def _copy_rowwise(param: torch.nn.Parameter, tensor_value, is_bias, rank, world_size):
+    """
+    This function copies the correct shard of the weights for a rowwise-TP'd module
+    according to the rank of the process and the world_size.
+
+    Args
+    ====
+    param: torch.nn.Parameter
+        Parameter that has had TP applied
+    tensor_value: torch.Tensor
+        tensor that needs sharding
+    rank: int
+        Rank of the current process
+    world_size: int
+        Total number of TP processes
+    """
+    # Divide the weight matrix along the last dimension.
+    if not is_bias:
+        output_size_per_partition = param.shape[1]
+        tensor = tensor_value[
+            :,
+            (rank * output_size_per_partition) : (
+                (rank + 1) * output_size_per_partition
+            ),
+        ]
+        param.copy_(tensor, non_blocking=True)
+    else:
+        if rank == 0:
+            _copy_if_present(param, tensor_value)
+        else:
+            param.zero_()
+
+
+def _copy_embedding(param: torch.nn.Parameter, tensor_value, rank, world_size):
+    """
+    This function copies the correct shard of the weights for a TP'd embedding module
+    according to the rank of the process and the world_size.
+
+    Args
+    ====
+    param: torch.nn.Parameter
+        Parameter that has had TP applied
+    tensor_value: torch.Tensor
+        tensor that needs sharding
+    rank: int
+        Rank of the current process
+    world_size: int
+        Total number of TP processes
+    """
+    # Divide the weight matrix along the last dimension.
+    output_size_per_partition = param.shape[1]
+    tensor = tensor_value[
+        :,
+        (rank * output_size_per_partition) : ((rank + 1) * output_size_per_partition),
+    ]
+    param.copy_(tensor, non_blocking=True)
+
+
+def _copy_if_present(parameter, tensor_value):
+    parameter.copy_(tensor_value, non_blocking=True)
+
+
+def _load_partial_state_dict(
+    model: torch.nn.Module,
+    state_dict,
+    tp_shard: bool,
+    rank=0,
+    world_size=1,
+):
+    unused_params = []
+    for key, tensor_value in state_dict.items():
```

Comment:
can we add some more comments in this loop about what is happening at each step

Reply:
done